### PR TITLE
FindStrict and FindAllStrict

### PR DESCRIPTION
**Pull request by Salmondx**
Implementation of `FindStrict` and `FindAllStrict` functions based on #15 discussion.
New test scenarios added for this functions and for the old `Find` and `FindAll`  functions.

Also I've implemented a little different algorithm for searching a value occurrence in a tag's attribute values.
In the previous implementation you were searching only for the first occurrence in attribute values:
```go
strings.Fields(n.Attr[i].Val)[0]  == args[2]
```
I've changed this behavior to search in all attribute values:
```go
func attributeContainsValue(attr html.Attribute, attribute, value string) bool {
	if attr.Key == attribute {
		for _, attrVal := range strings.Fields(attr.Val) {
			if attrVal == value {
				return true
			}
		}
	}
	return false
}
```
I think that in this question the order of the value doesn't matter, so there are no difference between `<div class="first second">` and `<div class="second first">` elements.

### CODE REVIEW

File: soup.go [hunk 1]
```diff
@@ -133,14 +132,44 @@ func (r Root) Find(args ...string) Root {
...
-	pointers := make([]Root, 0, 10)
+	pointers := make([]Root, 0, len(temp))
+	for i := 0; i < len(temp); i++ {
+		pointers = append(pointers, Root{temp[i], temp[i].Data, nil})
+	}
+	return pointers
+}
+
+// FindStrict finds the first occurrence of the given tag name
+// only if all provided attribute's values are exists
```
Comment (anaskhan96) [inline soup.go]: `// only if all the values of the provided attribute are an exact match` would be a better grammatical fit

File: soup.go [hunk 2]
```diff
@@ -133,14 +132,44 @@ func (r Root) Find(args ...string) Root {
...
+			panic("Element `" + args[0] + "` with attributes `" + strings.Join(args[1:], " ") + "` not found")
+		}
+		return Root{nil, "", errors.New("element `" + args[0] + "` with attributes `" + strings.Join(args[1:], " ") + "` not found")}
+	}
+	return Root{temp, temp.Data, nil}
+}
+
+// FindAllStrict finds all occurrences of the given tag name
+// only if all provided attribute's values are exists
+func (r Root) FindAllStrict(args ...string) []Root {
```
Comment (anaskhan96) [inline soup.go]: Again, `// only if all the values of the provided attribute are an exact match` would be a better grammatical fit

File: soup.go [hunk 3]
```diff
@@ -304,6 +341,21 @@ func findAllofem(n *html.Node, args []string) []*html.Node {
 	return nodeLinks
 }
 
+func attributeAndValueEquals(attr html.Attribute, attribute, value string) bool {
```
Comment (anaskhan96) [inline soup.go]: nit: short commented description of the function needed

File: soup.go [hunk 4]
```diff
@@ -304,6 +341,21 @@ func findAllofem(n *html.Node, args []string) []*html.Node {
...
+	return attr.Key == attribute && attr.Val == value
+}
+
+func attributeContainsValue(attr html.Attribute, attribute, value string) bool {
```
Comment (anaskhan96) [inline soup.go]: nit: short commented description of the function needed

**Comment by anaskhan96** [comment]
@Salmondx Thank you for the PR! The solution looks great. I've requested a few changes regarding the comments here and there, please go through them.

**Comment by Salmondx** [comment]
@anaskhan96 Ok, thanks for review, I'll fix this as soon as I can

**Comment by Salmondx** [comment]
@anaskhan96 Done

**Comment by anaskhan96** [comment]
Merged :)

**Pull request merged**